vault/identity_store_oidc_provider.go

```diff
@@ -95,12 +97,31 @@ type client struct {
 	Key            string        `json:"key"`
 	IDTokenTTL     time.Duration `json:"id_token_ttl"`
 	AccessTokenTTL time.Duration `json:"access_token_ttl"`
+	Type           ClientType    `json:"type"`
 
 	// Generated values that are used in OIDC endpoints
 	ClientID     string `json:"client_id"`
 	ClientSecret string `json:"client_secret"`
 }
 
+type ClientType int
```

Comment:
I've added the concept of client types to this PR. I went with an enum-style indication of the client type since it handles upgrades easily (default is 0, which maps to confidential via the iota). I explored the option of using just a string field for client type, but the upgrade scenario would've required special coding. I also explored the option of a simple public boolean value, but it wouldn't allow us to add more client types in the future (looking forward at OAuth 2.1 where there are [3 client types](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-04#section-2.1)).

Let me know your thoughts!

Reply:
I've done manual testing to ensure that upgrading results in existing clients to be confidential. I'm currently writing automated tests for the rest of the new behavior here.

Reply:
Tests added in 1bb4d1e, 54a23e0.